run_scripts/2_fetch_raw_data.py

```diff
@@ -70,43 +69,78 @@
+            raw_contacts = rapid_pro.get_raw_contacts(raw_export_log=f)
+
     # Download all the runs for each of the radio shows
-    for show in SHOWS:
-        output_file_path = f"{root_data_dir}/Raw Data/{show}.json"
-        print(f"Exporting show '{show}' to '{output_file_path}'...")
-
-        exit_code = subprocess.call([
-            "./docker-run.sh",
-            "--flow-name", show,
-            "--test-contacts-path", TEST_CONTACTS_PATH,
-            rapid_pro_domain,
-            rapid_pro_token,
-            user,
-            "all",
-            uuid_table_path,
-            output_file_path
-        ], cwd=f"{rapid_pro_tools_dir}/fetch_runs")
-
-        assert exit_code == 0, f"{rapid_pro_tools_dir}/fetch_runs/fetch_runs.py failed with exit_code {exit_code}"
-
-    # Download all the runs for each of the surveys
-    for survey in SURVEYS:
-        output_file_path = f"{root_data_dir}/Raw Data/{survey}.json"
-        print(f"Exporting survey '{survey}' to '{output_file_path}'...")
-
-        exit_code = subprocess.call([
-            "./docker-run.sh",
-            "--flow-name", survey,
-            "--test-contacts-path", TEST_CONTACTS_PATH,
-            rapid_pro_domain,
-            rapid_pro_token,
-            user,
-            "latest-only",
-            uuid_table_path,
-            output_file_path
-        ], cwd=f"{rapid_pro_tools_dir}/fetch_runs")
-
-        assert exit_code == 0, f"{rapid_pro_tools_dir}/fetch_runs/fetch_runs.py failed with exit_code {exit_code}"
+    for flow in SHOWS + SURVEYS:
+        runs_log_path = f"{root_data_dir}/Raw Data/{flow}_log.jsonl"
+        raw_runs_path = f"{root_data_dir}/Raw Data/{flow}_raw.json"
+        traced_runs_output_path = f"{root_data_dir}/Raw Data/{flow}.json"
+        print(f"Exporting show '{flow}' to '{traced_runs_output_path}'...")
+
+        flow_id = rapid_pro.get_flow_id(flow)
+
+        # Load the previous export of runs for this flow, and update them with the newest runs.
+        # If there is no previous export for this flow, fetch all the runs from Rapid Pro.
+        with open(runs_log_path, "a") as raw_export_log_file:
+            try:
+                print(f"Loading raw runs from file '{raw_runs_path}'...")
+                with open(raw_runs_path) as f:
+                    raw_runs = [Run.deserialize(run_json) for run_json in json.load(f)]
+                print(f"Loaded {len(raw_runs)} runs")
+                raw_runs = rapid_pro.update_raw_runs_with_latest_modified(
+                    flow_id, raw_runs, raw_export_log=raw_export_log_file)
+            except FileNotFoundError:
+                print(f"File '{raw_runs_path}' not found, will fetch all runs from the Rapid Pro server for flow '{flow}'")
+                raw_runs = rapid_pro.get_raw_runs_for_flow_id(flow_id, raw_export_log=raw_export_log_file)
+
+        # Fetch the latest contacts from Rapid Pro.
+        with open(contacts_log_path, "a") as f:
```

Comment:
I'm a little wary about reusing the file identifiers in Python like this. It's scoping rules are rather difficult to remember compared to other languages, so we don't get as much protection against accidentally reusing `f` as we would elsewhere. Normally that'll just cause an unpleasantly difficult to find bug, but when they're file identifiers it can cause riskier side effects, e.g. accidentally trampling on the logs.

Reply:
consider `contacts_log_file` instead

Reply:
changed